src/persist/src/s3.rs

```diff
@@ -270,8 +281,167 @@ impl S3Blob {
+        tracing::debug!(
+            "s3 PutObject single done {}b / {:?}",
+            value_len,
+            start_overall.elapsed()
+        );
+        Ok(())
+    }
+
+    async fn set_multi_part(&self, key: &str, value: Vec<u8>) -> Result<(), Error> {
+        let client = self.core.ensure_open()?;
+        let path = self.core.get_path(key);
+
+        let value = Bytes::from(value);
+        let start_overall = Instant::now();
+
+        // Start the multi part request and get an upload id.
+        tracing::trace!("s3 PutObject multi start {}b", value.len());
+        let upload_res = client
+            .create_multipart_upload()
+            .bucket(&self.core.bucket)
+            .key(&path)
+            .send()
+            .await
+            .map_err(|err| Error::from(format!("create_multipart_upload err: {}", err)))?;
+        let upload_id = upload_res.upload_id().ok_or_else(|| {
+            Error::from(format!(
+                "create_multipart_upload response missing upload_id"
+            ))
+        })?;
+        tracing::trace!(
+            "s3 create_multipart_upload took {:?}",
+            start_overall.elapsed()
+        );
+
+        // TODO: Plumb an Arc<AsyncRuntime> on S3BlobCore instead. I tried but
+        // was getting a "Cannot drop a runtime in a context where blocking is
+        // not allowed" error that I didn't want to get distracted with. All the
+        // calls into the s3 library require that this context is set anyway, so
+        // I suppose this is no worse than where we started.
+        let async_runtime = AsyncHandle::try_current().map_err(|err| err.to_string())?;
+
+        // Fire off all the individual parts.
+        //
+        // TODO: The aws cli throttles how many of these are outstanding at any
+        // given point. We'll likely want to do the same at some point.
+        let start_parts = Instant::now();
+        let mut part_futs = Vec::new();
+        for (part_num, part_range) in self.core.multipart_config.part_iter(value.len()) {
+            // NB: Without this spawn, these will execute serially. This is rust
+            // async 101 stuff, but there isn't much async in the persist
+            // codebase (yet?) so I thought it worth calling out.
+            let part_fut = async_runtime.spawn_named(
+                || "persist_s3blob_multipart",
+                client
+                    .upload_part()
+                    .bucket(&self.core.bucket)
+                    .key(&path)
+                    .upload_id(upload_id)
+                    .part_number(part_num as i32)
+                    .body(ByteStream::from(value.slice(part_range)))
+                    .send()
+                    .map(move |res| (start_parts.elapsed(), res)),
+            );
+            part_futs.push((part_num, part_fut));
+        }
+        let parts_len = part_futs.len();
+
+        // Wait on all the parts to finish. This is done in part order, no need
```

Comment:
Are we sure this is actually how it works? I'm testing a thing, give me just a second.

Reply:
Yep, it works. Though I'm still sometimes unsure with this async business.

Reply:
expanding a bit on my comment: you wouldn't need to spawn tasks if you used a future combinator on the upload futures, sth like `futures::future::join_all()`. This will automatically cancel all futures when one of them fails, so the task might give more control.

Overall, it's hard for me to say whether spawning tasks or using a combinator is better. Did you think about that?

Reply:
:+1: